internal/clients/tgw.go

```diff
@@ -0,0 +1,56 @@
+package clients
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/hashicorp/hcp-sdk-go/clients/cloud-network/preview/2020-09-07/client/network_service"
+	networkmodels "github.com/hashicorp/hcp-sdk-go/clients/cloud-network/preview/2020-09-07/models"
+	sharedmodels "github.com/hashicorp/hcp-sdk-go/clients/cloud-shared/v1/models"
+)
+
+// GetTGWAttachmentByID gets a TGW attachment by its ID, hvnID, and location
+func GetTGWAttachmentByID(ctx context.Context, client *Client, tgwAttachmentID string, hvnID string, loc *sharedmodels.HashicorpCloudLocationLocation) (*networkmodels.HashicorpCloudNetwork20200907TGWAttachment, error) {
+	getTGWAttachmentParams := network_service.NewGetTGWAttachmentParams()
+	getTGWAttachmentParams.ID = tgwAttachmentID
+	getTGWAttachmentParams.HvnID = hvnID
+	getTGWAttachmentParams.HvnLocationOrganizationID = loc.OrganizationID
+	getTGWAttachmentParams.HvnLocationProjectID = loc.ProjectID
+	getTGWAttachmentResponse, err := client.Network.GetTGWAttachment(getTGWAttachmentParams, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return getTGWAttachmentResponse.Payload.TgwAttachment, nil
+}
+
+// WaitForTGWAttachmentState will poll the GET TGW attachment endpoint until
+// it is in the specified state, ctx is canceled, or an error occurs. This
+// is required because AWS can return a newly created TGW attachment before
+// it is ready to be used.
+func WaitForTGWAttachmentState(ctx context.Context, client *Client, tgwAttachmentID string, hvnID string, loc *sharedmodels.HashicorpCloudLocationLocation, state string) (*networkmodels.HashicorpCloudNetwork20200907TGWAttachment, error) {
+	for {
+		tgwAtt, err := GetTGWAttachmentByID(ctx, client, tgwAttachmentID, hvnID, loc)
+		if err != nil {
+			return nil, fmt.Errorf("unable to retrieve Transit gateway attachment (%s): %v", tgwAttachmentID, err)
+		}
+		switch curState := string(tgwAtt.State); curState {
+		case state:
+			return tgwAtt, nil
+		case string(networkmodels.HashicorpCloudNetwork20200907TGWAttachmentStateFAILED):
```

Comment:
Are there any other terminal states to be aware of in regards to TGW?

Reply:
I'm not 100% sure, but I can check and make sure we're returning once we hit any terminal state. My guess is that we need to include a few more here.

Reply:
technically there're also `REJECTED` and `EXPIRED`, we might add it here too

Reply:
Other terminal states in addition to `FAILED` include `REJECTED`, `DELETING`, `EXPIRED`. 
`UNSET` being returned should also indicate something very wrong as well, however this should never happen, we are probably ok not handling this case here.

Reply:
Thanks for the additional context here (the diagram was really helpful too). I've updated the function to actually specify the "pending" states, instead of terminal states, to align with [the helper that the provider SDK offers](https://github.com/hashicorp/terraform-plugin-sdk/blob/master/helper/resource/state.go). This helper is used frequently in other providers, and seems to be a safer approach if we introduce new states in the future.